### PR TITLE
Add lock shard docs

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR adds internal docs for LockShard functionality.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x]  If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by frouioui — APPROVED** [review]
Should we also add this to the website repository?

File: doc/design-docs/LockShard.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+There are 2 variants of shard locking, `LockShard` which is a blocking call, and `TryLockShard` which tries to be a non-blocking call, but does not guarantee it.
+
+`TryLockShard` tries to find out if the shard is available to be locked or not. If it finds that the shard is locked, it returns with an error. However, there is still a race when the shard is not locked, that can cause `TryLockShard` to still block.
+
+### Working of LockShard
+
+`getLockTimeout` gets the amount of time we have to acquire a shard lock. It is not the amount of time that we acquire the shard lock for. It is currently misadvertised. LockShard returns a context, but that context doesn't have a timeout on it. When the shard lock expires, the context doesn't expire, because it doesn't have a timeout. To check whether the shard is locked or not, we have `CheckShardLocked`.
+
+The implementations of LockShard and CheckShardLocked differ slightly for all the different topology servers. We'll look at each of them separately.
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply doc/design-docs/LockShard.md]: Fixed

File: doc/design-docs/LockShard.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+This doc describes the working of shard locking that Vitess does using the topo servers.
+
+There are 2 variants of shard locking, `LockShard` which is a blocking call, and `TryLockShard` which tries to be a non-blocking call, but does not guarantee it.
+
+`TryLockShard` tries to find out if the shard is available to be locked or not. If it finds that the shard is locked, it returns with an error. However, there is still a race when the shard is not locked, that can cause `TryLockShard` to still block.
+
+### Working of LockShard
+
+`getLockTimeout` gets the amount of time we have to acquire a shard lock. It is not the amount of time that we acquire the shard lock for. It is currently misadvertised. LockShard returns a context, but that context doesn't have a timeout on it. When the shard lock expires, the context doesn't expire, because it doesn't have a timeout. To check whether the shard is locked or not, we have `CheckShardLocked`.
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply doc/design-docs/LockShard.md]: Fixed

**Review by rohit-nayak-ps — APPROVED** [review]

**Comment by rohit-nayak-ps** [comment]
> Should we also add this to the website repository?

This is implementation-related,  so it makes sense to keep it in the vitess repo I think.

**Comment by GuptaManan100** [comment]
I agree with Rohit, this can remain an internal Vitess doc

**Pull request merged**